### PR TITLE
Update package infrastructure

### PR DESCRIPTION
**Pull request by larrybradley**
This PR updates the package infrastructure.  Instead of breaking this up into many small PR, I did this all at once -- all of the changes are interrelated (sorry this PR overall is large - but each commit is atomic).

- Moved package metadata from `setup.cfg` to `pyproject.toml`.  Removed `setup.cfg` and `setup.py` files.
- Added the missing license file
- Updated the tox config, including NumPy 2.0dev and Python 3.12-dev tests
- Updated the CI workflows
- Fixed various test warnings (revealed by turning warnings into exceptions) 
- Updated Sphinx config (and removed its use of `pkg_resources`)
- Fixed the bandit config (which reveals issues that need to be fixed)

There will be known test failures with this PR.  

The devdeps test fails due to upstream NumPy 2.0dev deprecations in `stsci.tools`.  There are also upstream deprecations in `drizzlepac` and `matplotlib` that I've ignored for now.  Those can be removed with new upstream package releases.

~The bandit failures will need to be addressed in a separate PR.  I'll open an issue.~

Fixes: #46

### CODE REVIEW

**Comment by larrybradley** [comment]
I've temporarily allowed the bandit test to fail so that the other tests will run.

File: .flake8
Comment (pllim) [inline .flake8:1]: No ruff? I heard it's all the rage now. 😜

File: .github/workflows/ci_cron_daily.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        include:
+          - os: ubuntu-latest
+            python: '3.11'
+            tox_env: 'linkcheck'
+            allow_failure: false
+            prefix: ''
+
+          - os: ubuntu-latest
+            python: '3.11'
+            tox_env: 'py311-test-devdeps'
```
Comment (pllim) [inline .github/workflows/ci_cron_daily.yml:45]: Do you really need two flavors of devdeps here?

**Comment by larrybradley** [comment]
Followup tickets:

- #46 
- #47

File: .github/workflows/ci_cron_daily.yml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      uses: actions/checkout@v4
+      with:
+        fetch-depth: 0
+    - name: Set up Python ${{ matrix.python }}
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python }}
+    - name: Install base dependencies
+      run: python -m pip install --upgrade pip setuptools tox
+    - name: Print Python, pip, setuptools, and tox versions
```
Comment (pllim) [inline .github/workflows/ci_cron_daily.yml:68]: Why not use `pip freeze`?

File: .github/workflows/ci_tests.yml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            tox_env: 'py310-test'
+            allow_failure: false
+            prefix: ''
+
+          - os: ubuntu-latest
+            python: '3.11'
+            tox_env: 'py311-test-cov'
+            toxposargs: --remote-data=any
+            allow_failure: false
+            prefix: '(Temp allowed failure)'
```
Comment (pllim) [inline .github/workflows/ci_tests.yml]: These two lines are not consistent with one another.

File: .github/workflows/ci_tests.yml [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+          - os: ubuntu-latest
+            python: '3.11'
+            tox_env: 'securityaudit'
+            allow_failure: true
+            prefix: '(Temp allowed failure)'
+
+          - os: ubuntu-latest
+            python: '3.11'
+            tox_env: 'py311-test-devdeps'
```
Comment (pllim) [inline .github/workflows/ci_tests.yml:90]: Do you need two flavors of devdeps here?

File: .github/workflows/ci_cron_daily.yml
Comment (pllim) [inline .github/workflows/ci_cron_daily.yml:1]: Do you really need anything besides linkcheck here given that devdeps is already in `ci_tests.yml` and that file also has a cron schedule?

File: .github/workflows/ci_tests.yml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      uses: actions/checkout@v4
+      with:
+        fetch-depth: 0
+    - name: Set up Python ${{ matrix.python }}
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python }}
+    - name: Install base dependencies
+      run: python -m pip install --upgrade pip setuptools tox
+    - name: Print Python, pip, setuptools, and tox versions
```
Comment (pllim) [inline .github/workflows/ci_tests.yml:113]: Why not use `pip freeze`?

**Review by pllim — COMMENTED** [review]
I didn't look too carefully but here are some comments.

File: .github/workflows/ci_cron_daily.yml (hunk 1 above)
Comment (larrybradley) [reply .github/workflows/ci_cron_daily.yml:45]: One specifically tests Python 3.12-dev.

File: .flake8
Comment (larrybradley) [reply .flake8:1]: grumble, grumble.

File: .github/workflows/ci_tests.yml (hunk 3 above)
Comment (larrybradley) [reply .github/workflows/ci_tests.yml]: Fixed

File: .github/workflows/ci_tests.yml (hunk 4 above)
Comment (larrybradley) [reply .github/workflows/ci_tests.yml:90]: one tests py312-dev

File: docs/index.rst [hunk 6]
```diff
@@ -12,17 +12,17 @@ Slitlessutils
...
 :Developers:
    Russell Ryan, Gagandeep Anand, Roberto Avila, Larry Bradley, Norman Grogin, Nimish Hathi, Hatice Karatay, Benjamin Kuhn, Duy Nguyen, Richard O'Steen, Camila Pacifici, Aidan Pidgeon, Lee Quick, Debopam Som, Megan Sosey
 
 
 
-   
 
 
-   
+
```
[the remark this answers is not in the document]
Comment (larrybradley) [reply docs/index.rst:24]: My editor trims trailing whitespace

File: .github/workflows/ci_cron_daily.yml (hunk 1 above)
Comment (pllim) [reply .github/workflows/ci_cron_daily.yml:45]: I still don't see why you need to separate it out like this, but ok.

File: pyproject.toml [hunk 7]
```diff
@@ -1,47 +1,148 @@
...
-
 [project]
-name = "slitlessutils"
+name = 'slitlessutils'
+description = 'A compendium of tools to analyze slitless spectroscopy'
+readme = 'README.md'
+license = {file = 'LICENSE.rst'}
 authors = [
-    {name = "Russell Ryan", email = "rryan@stsci.edu"},
+    {name = 'Russell Ryan', email = 'rryan@stsci.edu'},
```
[the remark this answers is not in the document]
Comment (larrybradley) [reply pyproject.toml:7]: This should probably be changed to something like "slitlessutils developers" and "help@stsci.edu".
@Russell-Ryan is that OK?

File: pyproject.toml [hunk 8]
```diff
@@ -1,47 +1,148 @@
...
+    'sphinx-astropy>=1.6',
+    'sphinx-rtd-theme',
+    'tomli; python_version < "3.11"',
 ]
+
+[build-system]
+requires = [
+    'setuptools>=61.2',
+    'setuptools-scm>=6.2',
+    'numpy>=1.22.3'
```
[the remark this answers is not in the document]
Comment (larrybradley) [reply pyproject.toml:71]: This package is numpy 2 compatible.  But some dependencies (e.g., stsci.tools) are not.

File: pyproject.toml [hunk 9]
```diff
@@ -1,47 +1,148 @@
...
+    'ignore:unclosed file:ResourceWarning',
+    # remove after drizzlepac is updated
+    'ignore: `photutils.DAOStarFinder` is a deprecated alias:DeprecationWarning',
+    # deprecation warning from stsci.tools via drizzlepac;
+    # remove after new stsci.tools release
+    'ignore: distutils Version classes are deprecated. Use packaging.version instead.:DeprecationWarning',
+
+    'ignore: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.:UserWarning',
+    # python 3.12 deprecation in matplotlib 3.9dev
+    'ignore:datetime.datetime.utcfromtimestamp\(\) is deprecated:DeprecationWarning',
```
[the remark this answers is not in the document]
Comment (larrybradley) [reply pyproject.toml:115]: I was still getting this warning from mpl dev

[the remark this answers is not in the document]
Comment (pllim) [reply pyproject.toml:115]: Oh, hmm. Maybe I misremember then. You can ignore.

File: pyproject.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (pllim) [reply pyproject.toml:71]: Do we need a new `stsci.tools` release with https://github.com/spacetelescope/stsci.tools/pull/153 ?

[the remark this answers is not in the document]
Comment (larrybradley) [reply pyproject.toml:71]: Yes!

File: .github/workflows/ci_cron_daily.yml (hunk 1 above)
Comment (larrybradley) [reply .github/workflows/ci_cron_daily.yml:45]: I didn't find any Python 3.12 issues, so perhaps I'll just use py 3.12 for the devdeps.  Is that what you are using?

File: pyproject.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (Russell-Ryan) [reply pyproject.toml:7]: fine with me.  I didn't make that.

File: pyproject.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (pllim) [reply pyproject.toml:71]: Okay I released https://pypi.org/project/stsci.tools/4.1.0 . Hopefully it works. Its test suite coverage isn't great. If you find any remaining incompatibility, please open issue. Thanks!

File: .github/workflows/ci_cron_daily.yml (hunk 2 above)
Comment (larrybradley) [reply .github/workflows/ci_cron_daily.yml:68]: `pip freeze` doesn't include those packages, e.g.:
https://github.com/spacetelescope/slitlessutils/actions/runs/6317086199/job/17153095889#step:6:24

File: .github/workflows/ci_cron_daily.yml (hunk 1 above)
Comment (larrybradley) [reply .github/workflows/ci_cron_daily.yml:45]: I switched to one devdeps with Python 3.12.

File: pyproject.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (larrybradley) [reply pyproject.toml]: Done, thanks.

File: pyproject.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (larrybradley) [reply pyproject.toml:71]: I removed the ignore for that warning.  🤞

File: pyproject.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (larrybradley) [reply pyproject.toml]: I removed it, let's see.  🤞

File: .github/workflows/ci_cron_daily.yml
Comment (larrybradley) [reply .github/workflows/ci_cron_daily.yml:1]: The `ci_tests` cron job runs weekly.  This runs daily (except for the day the `ci_test` cron job runs).  So many thinks are breaking recently. :(

File: pyproject.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (larrybradley) [reply pyproject.toml:71]: Ah, I misunderstood your comment at first.  I didn't realize this package had pinned numpy.  Indeed, it did not work with >= 1.25, but I just fixed that.

**Comment by larrybradley** [comment]
@pllim Thanks for the reviews!

**Pull request merged**